libutils/file_lib.c

```diff
@@ -467,6 +467,94 @@ void switch_symlink_hook();
 #define TEST_SYMLINK_SWITCH_POINT
 #endif
 
+void PathWalk(
+    const char *const path,
+    PathWalkFn callback,
+    void *const data,
+    void *(*copy)(void *),
+    void (*destroy)(void *))
+{
+    assert(path != NULL);
+    assert(callback != NULL);
+
+    Seq *const children = ListDir(path, NULL);
+    if (children == NULL) {
+        Log(LOG_LEVEL_DEBUG, "Failed to list directory '%s'", path);
+        return;
+    }
+    const size_t n_children = SeqLength(children);
+
+    Seq *const dirnames = SeqNew(1, free);
+    Seq *const filenames = SeqNew(1, free);
+
+    for (size_t i = 0; i < n_children; i++)
+    {
+        /* The basename(3) function might potentially mutate the child, but we
+         * don't mind. */
+        char *const child = SeqAt(children, i);
+        const char *const b_name = basename(child);
+
+        /* We don't iterate the '.' and '..' directory entries as it would cause
+         * infinite recursion. */
+        if (StringEqual(b_name, ".") || StringEqual(b_name, ".."))
+        {
+            continue;
+        }
+
+        struct stat sb;
+        if (stat(child, &sb) == 0)
```

Comment:
Note that `stat()` follows symlinks.

Reply:
Do you think we should not follow symlinks?